### PR TITLE
enhance: detect config / client compatibility mismatch in cli

### PR DESCRIPTION
**Pull request by hikchoi**
# enhance: detect config / client compatibility mismatch in cli
This PR:
- adds compatibility validation for config and client.
- adds validation logic to CLI commands
  - all cli commands excluding `dendron dev` will have have this validation logic running prior to execution.
  - when there is a client / config compatibility mismatch, the process will exit and display a message informing how to fix it.
  - image
  - image
  - later this will be enhanced to have an option to pass `--yes` to automatically fix (if possible) the mismatch (useful in CI/CD)
- adds tracking of client / config mismatch events
- fixes client version fetching logic in CLI (will now fetch from package.json rather than `process.env.npm_package_version` which will alway be `undefined` because we are not running an npm script.)

NOTE
- splitting some changes that will be more immediately useful from this task: [[Config Migration Implementation|dendron://private/task.workspace.2022.01.03.config-migration-implementation]] in the spirit of iteration / timely delivery.

# Pull Request Checklist

If this is your first time submitting a pull request to Dendron, copy and paste the full [Dendron Review Checklist](https://docs.dendron.so/notes/1EoNIXzgmhgagqcAo9nDn.html) into this request and check off each item as necessary. 

This template contains the short checklist which is used by the Dendron core team. 

## Testing
- [x] [Write Tests](https://docs.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html#writing-tests) 
- [x] [Confirm existing tests pass](https://docs.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html#executing-tests)
- [x] [Confirm manual testing](https://docs.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html#manual-testing) 
- [ ] If your tests changes an existing snapshot, make sure that snapshots are [updated](https://docs.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html#updating-test-snapshots)
- [ ] If you are adding a new language feature (graphically visible in VS Code/preview/publishing), make sure that it is included in the [test workspace](https://docs.dendron.so/notes/dtMsF12SF2SUhLN10sYe2.html)
### Docs
- [x] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR
### Analytics
- [x] if you are adding analytics related changes, make sure the [Telemetry](https://docs.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated

### CODE REVIEW

**Comment by hikchoi** [comment]
just rebased in case the recent bunch of refactoring in the codebase affects the features added here.

**Review by jonathanyeung — APPROVED** [review]

**Comment by jonathanyeung** [comment]
minor nit: I find one aspect of the console message a little bit confusing. 

- Maybe when config is out of date, let's not show the line "minimum compatible client version" (and just the other 3 lines)
- When Client is out of date, let's not show "minimum compatible config version"?

Just a minor nit, otherwise the console prompt is very helpful.

File: packages/dendron-cli/src/commands/base.ts [hunk 1]
```diff
@@ -102,6 +106,51 @@ export abstract class CLICommand<
...
+        `${reason} is out of date.`,
+        "------------------------------------------------",
+        `current client version:            ${clientVersionOkay} ${clientVersion}`,
+        `minimum compatible client version:    ${minCompatClientVersion}`,
+        `current config version:            ${configVersionOkay} ${configVersion}`,
+        `minimum compatible config version:    ${minCompatConfigVersion}`,
+        "------------------------------------------------",
+        instruction,
+      ].join("\n");
+      console.log(message);
```
Comment (kevinslin) [inline packages/dendron-cli/src/commands/base.ts:142]: can you add eslint disable next line?

File: packages/dendron-cli/src/commands/base.ts [hunk 2]
```diff
@@ -102,6 +106,51 @@ export abstract class CLICommand<
     this.L.info({ msg: `Telemetry is disabled? ${segment.hasOptedOut}` });
   }
 
+  async validateConfig(opts: { wsRoot: string }) {
+    const { wsRoot } = opts;
+    const ws = new WorkspaceService({ wsRoot });
+    const dendronConfig = ws.config;
+    const configVersion = ConfigUtils.getProp(dendronConfig, "version");
```
Comment (kevinslin) [inline packages/dendron-cli/src/commands/base.ts:113]: if the `version` doesn't exist, we currently assume that it equals 4. can this get us into trouble?

File: packages/dendron-cli/src/commands/base.ts [hunk 3]
```diff
@@ -102,6 +106,51 @@ export abstract class CLICommand<
...
+    const clientVersion = CLIUtils.getClientVersion();
+    const validationResp = ConfigUtils.configIsValid({
```
Comment (kevinslin) [inline packages/dendron-cli/src/commands/base.ts:115]: its okay to say this is out of scope for this but we should also add a followup to check if the config is **actually** valid for the given version.
a user can keep the verison but manually change the settings. ideally, we shhould use json schema to validate this (lets not use a mess of if statements).
each config version can come with its own json schema and we can use `ajv` to validate

**Review by kevinslin — CHANGES_REQUESTED** [review]

File: packages/dendron-cli/src/commands/base.ts (hunk 3 above)
Comment (hikchoi) [reply packages/dendron-cli/src/commands/base.ts:115]: yes this is the plan down the line.

File: packages/dendron-cli/src/commands/base.ts (hunk 2 above)
Comment (hikchoi) [reply packages/dendron-cli/src/commands/base.ts:113]: That's right. would need to look at the raw value and not assume a default with `getProp`.
For now I think it makes sense to stop all execution if the `version` field doesn't exist.
Later down the line once we introduce json schema validation, we will have to take snapshots of every version of the JSON schema and use that to determine the real `version`

**Comment by hikchoi** [comment]
UPDATE:

1. made the message less confusing:

image

image

2. will now display a message to make sure `version` is present and correct in `dendron.yml`
3. now use `this.print` instead of bare `console.log`

**Review by kevinslin — APPROVED** [review]

**Pull request merged**